### PR TITLE
Feature/serializer

### PR DESCRIPTION
**Pull request by basz**

### CODE REVIEW

File: docs/snapshots.md [hunk 1]
```diff
@@ -21,6 +21,34 @@ aggregate loading.
...
+```
+
+If you are using the interop factories all you have to do is create a Factory for `Prooph\SnapshotStore\Serializer` and add that as dependency;
+
+```
+<?php
+
+return [
+	'dependencies' => [
+		'factories' => \Prooph\SnapshotStore\Serializer::class => My\CallbackSerializerFactory::class,
```
Comment (prolic) [inline docs/snapshots.md]: `[` missing

File: src/CallbackSerializer.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+     * callable
+     */
+    private $serializer = 'serialize';
+
+    /**
+     * callable
+     */
+    private $deserialize = 'unserialize';
+
+    public function __construct(?callable $serializer, ?callable $deserialize)
```
Comment (prolic) [inline src/CallbackSerializer.php]: use `$unserialize` instead of `$deserialize` please.

File: src/CallbackSerializer.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+{
+    /**
+     * callable
+     */
+    private $serializer = 'serialize';
+
+    /**
+     * callable
+     */
+    private $deserialize = 'unserialize';
```
Comment (prolic) [inline src/CallbackSerializer.php]: private $unserialize

Comment (prolic) [reply src/CallbackSerializer.php]: or maybe better $unserializeCallback

File: src/CallbackSerializer.php (hunk 2 above)
Comment (prolic) [reply src/CallbackSerializer.php]: maybe better `?callable $serializeCallback, ?callable $unserializeCallback`

File: tests/CallbackSerializerTest.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        $after = $serializer->unserialize($serialized);
+
+        $this->assertEquals('O:8:"stdClass":0:{}', $serialized);
+        $this->assertEquals($before, $after);
+    }
+
+    /**
+     * @test
+     */
+    public function it_both_callback_are_required_before_default_is_not_used(): void
```
Comment (prolic) [inline tests/CallbackSerializerTest.php]: method name: `it_uses_default_if_only_one_callback_provided_instead_of_two`

**Review by prolic — CHANGES_REQUESTED** [review]

File: docs/snapshots.md [hunk 5]
```diff
@@ -21,6 +21,35 @@ aggregate loading.
 
 Our example application [proophessor-do](https://github.com/prooph/proophessor-do) contains a snapshotting tutorial.
 
+## Using a different Serializer. 
+
+By default we use php build in serialize/unserialize methods to around aggregate persistence. If you would like to change it you can do that since v1.1.
```
Comment (bweston92) [inline docs/snapshots.md]: By default prooph uses PHP's own serialise and unserialize methods. These may not suite your needs so as of v1.1 of the snapshot store you can now use a custom serialiser.

File: src/CallbackSerializer.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    private $serializeCallback = 'serialize';
+
+    /**
+     * callable
+     */
+    private $unserializeCallback = 'unserialize';
+
+    public function __construct(?callable $serializeCallback, ?callable $unserializeCallback)
+    {
+        if (null !== $serializeCallback && null !== $unserializeCallback) {
```
Comment (bweston92) [inline src/CallbackSerializer.php:29]: Would it not be better allow custom on either one? For example I might want only want to change something before I serialize. Or I may want to just update a time stamp when I unserialise.

File: src/CallbackSerializer.php [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        }
+    }
+
+    /**
+     * @param object|array $data
+     * @return string
+     */
+    public function serialize($data): string
+    {
+        return call_user_func_array($this->serializeCallback, [$data]);
```
Comment (bweston92) [inline src/CallbackSerializer.php]: call_user_func_array is for dynamic length arguments. Use call_user_func as it is faster.

Example: return call_user_func($this->serializeCallback, $data);

File: src/CallbackSerializer.php [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    /**
+     * @param string $serialized
+     * @return object|array
+     */
+    public function unserialize(string $serialized)
+    {
+        return call_user_func_array($this->unserializeCallback, [$serialized]);
```
Comment (bweston92) [inline src/CallbackSerializer.php]: call_user_func_array is for dynamic length arguments. Use call_user_func as it is faster.

Example: return call_user_func($this->in serializeCallback, $serialized);

**Review by bweston92 — COMMENTED** [review]
Please review my concerns.

**Comment by bweston92** [comment]
Good work @basz just a few things to pick up on

File: src/CallbackSerializer.php (hunk 6 above)
Comment (basz) [reply src/CallbackSerializer.php:29]: don't you think we should discourage such behaviour?

Changing anything on the aggregate should not be done at the snapshot layer... When a snapshot does not exists or isn't up to date events are going to reconstitute it.

This check does not makes it *impossible* but atleast one might be more aware of that. What goes in comes out. Unless i'm missing something?

File: src/CallbackSerializer.php (hunk 7 above)
Comment (basz) [reply src/CallbackSerializer.php]: cool then this should be faster too (not here of course)
```
call_user_func($callable, ...$someArray)
```

File: src/CallbackSerializer.php (hunk 6 above)
Comment (bweston92) [reply src/CallbackSerializer.php:29]: Nope you're not missing anything, that is most certainly true it should be in exactly the same state.

File: src/CallbackSerializer.php (hunk 7 above)
Comment (bweston92) [reply src/CallbackSerializer.php]: Would be an interesting benchmark. I only remember doing it for call_user_func and call_user_func_array without variadic params.

File: src/CallbackSerializer.php (hunk 6 above)
Comment (prolic) [reply src/CallbackSerializer.php:29]: I am against changing timestamp or anything else within the snapshot. If you want to do some crazy weird stuff, implement your own snapshot-store.

**Pull request merged**